src/desktop/main.cc

```diff
@@ -652,72 +567,159 @@ MAIN {
-      if (resolveWindow) {
-        resolveWindow->resolvePromise(message.get("seq"), OK_STATE, std::to_string(index));
+
+      if (!targetWindow || targetWindowStatus == WindowManager::WindowStatus::WINDOW_NONE) {
+        JSON::Object json = JSON::Object::Entries {
+          { "err", targetWindowStatus }
+        };
+        resolveWindow->resolvePromise(message.get("seq"), ERROR_STATE, json.str());
+        return;
       }
+
+      targetWindow->hide(EMPTY_SEQ);
+
+      JSON::Object json = JSON::Object::Entries {
+        { "data", targetWindow->json() }
+      };
+      resolveWindow->resolvePromise(message.get("seq"), OK_STATE, json.str());
+      return;
+    }
+
+    if (message.name == "window.setTitle") {
+      const auto seq = message.seq;
+      const auto currentIndex = message.index;
+      const auto currentWindow = windowManager.getWindow(currentIndex);
+      const auto targetWindowIndex = message.get("targetWindowIndex").size() > 0 ? std::stoi(message.get("targetWindowIndex")) : currentIndex;
+      const auto targetWindow = windowManager.getWindow(targetWindowIndex);
+
+      targetWindow->setTitle(value);
+
+      JSON::Object json = JSON::Object::Entries {
+        { "data", targetWindow->json() },
+      };
+      currentWindow->resolvePromise(seq, OK_STATE, json.str());
       return;
     }
 
     if (message.name == "window.navigate") {
-      auto targetWindowIndex = std::stoi(message.get("window"));
-      targetWindowIndex = targetWindowIndex < 0 ? 0 : targetWindowIndex;
-      auto index = message.index < 0 ? 0 : message.index;
-      auto window = windowManager.getWindow(targetWindowIndex);
-      auto url = message.get("url");
-      navigate(window, cwd, EMPTY_SEQ, decodeURIComponent(url));
+      const auto seq = message.seq;
+      const auto currentIndex = message.index;
+      const auto currentWindow = windowManager.getWindow(currentIndex);
+      const auto targetWindowIndex = message.get("targetWindowIndex").size() > 0 ? std::stoi(message.get("targetWindowIndex")) : currentIndex;
+      const auto targetWindow = windowManager.getWindow(targetWindowIndex);
+      const auto url = message.get("url");
+      const auto error = getNavigationError(cwd, decodeURIComponent(url));
 
-      auto resolveWindow = windowManager.getWindow(index);
-      if (resolveWindow) {
-        resolveWindow->resolvePromise(message.get("seq"), OK_STATE, std::to_string(index));
+      if (error.size() > 0) {
+        JSON::Object json = JSON::Object::Entries {
+          { "err", error }
+        };
+        currentWindow->resolvePromise(seq, ERROR_STATE, json.str());
+        return;
       }
+
+      targetWindow->navigate(seq, url);
+
+      JSON::Object json = JSON::Object::Entries {
+        { "data", targetWindow->json() },
+      };
+
+      currentWindow->resolvePromise(seq, OK_STATE, json.str());
+      return;
+    }
+
+    if (message.name == "window.showInspector") {
```

Comment:
does this work on linux and windows? cc @mribbons

Reply:
Inspector menu works on windows :+1: @chicoxyzzy @heapwolf